app/controllers/visualizations_controller.rb

```diff
@@ -214,7 +214,7 @@ def background_for_heatmap
 
   # The most recent update time of any samples pipeline run.
   def heatmap_ts
-    updated_ats = samples_for_heatmap.map { |sample| sample.first_pipeline_run.updated_at }
+    updated_ats = samples_for_heatmap.map { |sample| sample.first_pipeline_run && sample.first_pipeline_run.updated_at }
```

Comment:
There is an easier syntax for that:

`sample&.first_pipeline_run&.updated_at`

However I have a question: is it fine to return null for these cases? Shouldn't it be something like:

```
 updated_ats = samples_for_heatmap
                          .select { |sample| sample.first_pipeline_run } 
                          .map { |sample| sample.first_pipeline_run.updated_at }
```

Reply:
It's okay to return null because of the `max` below but it's probably better to filter with `select`.

Reply:
Actually, after reading Tiago's comments, I'm wondering if we should fix this query instead:

```
  def samples_for_heatmap
    id = visualization_params[:id]
    sample_ids = id ? Visualization.find(id).sample_ids : params[:sampleIds]
    current_power.samples
                 .where(id: sample_ids)
                 .includes([:host_genome, :pipeline_runs, metadata: [:metadata_field]])
  end
```

Maybe it shouldn't return samples that have no pipeline_runs.

Reply:
Samples that have no pipeline_runs are filtered out deeper inside the heatmap code.

Reply:
And is there any use for samples that have no pipeline_run in the heatmap? I'm asking that because it seems that fixing the query would simplify the overall solution

Reply:
As it happens, when I was working on heatmap perf, I found there are a small number of samples all dating to 2018 that have taxon counts but no pipeline runs. I'm not sure why that would be so I didn't want to change the logic.

Reply:
I was discussing with Tiago and actually we are puzzled how could a Sample has no pipeline runs. Ok, let's just fix the issue now. We can review this flow more in depth later on.

Reply:
I'd also love to know why!